File: backend/app/schemas/equivalence.py
"""Pydantic schemas for equivalence group endpoints."""

from datetime import datetime
from .common import UTCTimestamp

from pydantic import BaseModel, Field


# ═══════════════════════════════════════════════════════════════════════════════
# Request schemas
# ═══════════════════════════════════════════════════════════════════════════════


class EquivalenceGroupCreate(BaseModel):
    label: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    column_ids: list[int] = Field(default_factory=list)


class EquivalenceGroupBulkCreate(BaseModel):
    groups: list[EquivalenceGroupCreate]


class EquivalenceGroupUpdate(BaseModel):
    label: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = None


class EquivalenceGroupAddColumns(BaseModel):
    column_ids: list[int] = Field(..., min_length=1)


class EquivalenceGroupRemoveColumns(BaseModel):
    column_ids: list[int] = Field(..., min_length=1)


class EquivalenceGroupReorderRequest(BaseModel):
    group_ids: list[int]


class ColumnSwap(BaseModel):
    """A single pair of columns to swap equivalence_group assignments."""
    column_id_a: int
    column_id_b: int


class EquivalenceGroupSwapRequest(BaseModel):
    """Tier 3 crosswalk swap endpoint — atomically exchange each pair's
    equivalence_group_id. All pairs must pass validation before any swap
    is applied (all-or-nothing atomicity via the transaction).
    """
    swaps: list[ColumnSwap] = Field(..., min_length=1)


# ═══════════════════════════════════════════════════════════════════════════════
# Response schemas
# ═══════════════════════════════════════════════════════════════════════════════


class EquivalenceGroupColumnDefInfo(BaseModel):
    id: int
    name: str
    recode_type: str
    is_primary: bool


class EquivalenceGroupColumnInfo(BaseModel):
    id: int
    dataset_id: int
    dataset_name: str
    column_code: str | None = None
    column_text: str
    column_type: str
    scale_labels: list[str] | None = None
    scale_points: int | None = None
    recode_definitions: list[EquivalenceGroupColumnDefInfo] = []


class EquivalenceGroupResponse(BaseModel):
    id: int
    project_id: int
    label: str
    description: str | None = None
    origin: str
    columns: list[EquivalenceGroupColumnInfo]
    created_at: UTCTimestamp
    updated_at: UTCTimestamp


class EquivalenceGroupListResponse(BaseModel):
    groups: list[EquivalenceGroupResponse]
    total: int


class EquivalenceGroupSwapResponse(BaseModel):
    """Response: the groups affected by the swap (loaded after the swap
    completes), plus which metric IDs were recomputed synchronously per
    GAP 3.13 Option A.
    """
    updated_groups: list[EquivalenceGroupResponse]
    recomputed_metric_ids: list[int] = Field(default_factory=list)


class EquivalenceGroupRemoveColumnsResponse(BaseModel):
    """Response from `remove_columns`: either the updated group (if columns
    remain) or `dissolved=True` (if the removal emptied the group and the
    backend auto-deleted it).

    Path A (#323): empty equivalence groups are auto-dissolved server-side.
    The frontend reads `dissolved` directly to update its caches; previously
    the bandage in `useCrosswalkMutations.ts::removeColumnFromRowMutation`
    inferred this from `columns.length === 0` and fired a follow-up DELETE.
    """
    group: EquivalenceGroupResponse | None = None
    dissolved: bool = False


class BulkCreateResult(BaseModel):
    created: int
    groups: list[EquivalenceGroupResponse]


# ═══════════════════════════════════════════════════════════════════════════════
# Suggest schemas
# ═══════════════════════════════════════════════════════════════════════════════


class SuggestedGroupColumn(BaseModel):
    id: int
    dataset_id: int
    dataset_name: str
    column_code: str | None = None
    column_text: str
    column_type: str
    similarity_score: float | None = None


class SuggestedGroup(BaseModel):
    label: str
    match_type: str  # "exact_text", "code_match", or "similar_text"
    type_mismatch: bool
    similarity_score: float | None = None
    columns: list[SuggestedGroupColumn]


class EquivalenceSuggestResponse(BaseModel):
    suggestions: list[SuggestedGroup]


# ═══════════════════════════════════════════════════════════════════════════════
# All-columns schemas
# ═══════════════════════════════════════════════════════════════════════════════


# ═══════════════════════════════════════════════════════════════════════════════
# Find-matches schemas
# ═══════════════════════════════════════════════════════════════════════════════


class FindMatchesRequest(BaseModel):
    column_ids: list[int] = Field(..., min_length=1)
    min_similarity: float = Field(default=0.70, ge=0.0, le=1.0)


class ColumnMatchResult(BaseModel):
    anchor_column_id: int
    target_column_id: int
    target_column_text: str
    target_column_code: str | None = None
    target_dataset_id: int
    target_dataset_name: str
    target_column_type: str
    similarity: float
    already_linked: bool
    # True when the target is itself one of the user's explicitly selected anchor
    # columns (from a different dataset). These candidates bypass the fuzzy-match
    # threshold and are pre-checked in the UI.
    user_selected: bool = False


class FindMatchesResponse(BaseModel):
    matches: list[ColumnMatchResult]


# ═══════════════════════════════════════════════════════════════════════════════
# All-columns schemas
# ═══════════════════════════════════════════════════════════════════════════════


class ProjectColumnInfo(BaseModel):
    id: int
    dataset_id: int
    dataset_name: str
    # Denormalized dataset color (mirrors `dataset_name` denormalization) so
    # crosswalk surfaces resolve dataset visual identity without a second
    # query. Null when the dataset has no user-set color override.
    dataset_color: str | None = None
    column_code: str | None = None
    column_name: str | None = None
    column_text: str
    column_type: str
    scale_points: int | None = None
    # Phase 4.5: enrich the all-columns endpoint so the crosswalk can drive
    # the scale-labels mismatch icon (v2: full label-array comparison) and
    # the type picker's recode-defs pre-flight gate without a second query.
    scale_labels: list[str] | None = None
    recode_def_count: int = 0
    equivalence_group_id: int | None = None
    equivalence_group_label: str | None = None


class ProjectColumnListResponse(BaseModel):
    columns: list[ProjectColumnInfo]
    total: int

File: frontend/src/lib/api/equivalence.ts
import api from './client'

// Equivalence group types
export interface EquivalenceGroupColumnDefInfo {
  id: number
  name: string
  recode_type: string
  is_primary: boolean
}

export interface EquivalenceGroupColumnInfo {
  id: number
  dataset_id: number
  dataset_name: string
  column_code: string | null
  column_text: string
  column_type: string
  scale_labels: string[] | null
  scale_points: number | null
  recode_definitions: EquivalenceGroupColumnDefInfo[]
}

export interface EquivalenceGroupResponse {
  id: number
  project_id: number
  label: string
  description: string | null
  origin: string
  columns: EquivalenceGroupColumnInfo[]
  created_at: string
  updated_at: string
}

export interface EquivalenceGroupListResponse {
  groups: EquivalenceGroupResponse[]
  total: number
}

export interface BulkCreateResult {
  created: number
  groups: EquivalenceGroupResponse[]
}

export interface SuggestedGroupColumn {
  id: number
  dataset_id: number
  dataset_name: string
  column_code: string | null
  column_text: string
  column_type: string
  similarity_score: number | null
}

export interface SuggestedGroup {
  label: string
  match_type: 'exact_text' | 'code_match' | 'similar_text'
  type_mismatch: boolean
  similarity_score: number | null
  columns: SuggestedGroupColumn[]
}

export interface EquivalenceSuggestResponse {
  suggestions: SuggestedGroup[]
}

export interface ColumnMatchResult {
  anchor_column_id: number
  target_column_id: number
  target_column_text: string
  target_column_code: string | null
  target_dataset_id: number
  target_dataset_name: string
  target_column_type: string
  similarity: number
  already_linked: boolean
  /** True when the target is another anchor the user explicitly picked from a
   * different dataset. These bypass the fuzzy threshold and are pre-checked. */
  user_selected: boolean
}

export interface FindMatchesResponse {
  matches: ColumnMatchResult[]
}

export interface ColumnSwap {
  column_id_a: number
  column_id_b: number
}

export interface EquivalenceGroupSwapResponse {
  updated_groups: EquivalenceGroupResponse[]
  recomputed_metric_ids: number[]
}

/** Path A (#323): `remove_columns` returns the post-mutation group OR
 * `dissolved=true` when the removal emptied the group and the backend
 * auto-deleted it in the same transaction. The frontend reads `dissolved`
 * directly and patches caches accordingly. */
export interface EquivalenceGroupRemoveColumnsResponse {
  group: EquivalenceGroupResponse | null
  dissolved: boolean
}

// API functions - Equivalence Groups
export const equivalenceApi = {
  list: (projectId: number) =>
    api.get<EquivalenceGroupListResponse>(`/projects/${projectId}/equivalence-groups`).then(res => res.data),
  create: (projectId: number, data: { label: string; description?: string | null; column_ids?: number[] }) =>
    api.post<EquivalenceGroupResponse>(`/projects/${projectId}/equivalence-groups`, data).then(res => res.data),
  bulkCreate: (projectId: number, groups: Array<{ label: string; description?: string | null; column_ids: number[] }>) =>
    api.post<BulkCreateResult>(`/projects/${projectId}/equivalence-groups/bulk`, { groups }).then(res => res.data),
  update: (projectId: number, groupId: number, data: { label?: string; description?: string | null }) =>
    api.patch<EquivalenceGroupResponse>(`/projects/${projectId}/equivalence-groups/${groupId}`, data).then(res => res.data),
  delete: (projectId: number, groupId: number) =>
    api.delete(`/projects/${projectId}/equivalence-groups/${groupId}`).then(res => res.data),
  addColumns: (projectId: number, groupId: number, columnIds: number[]) =>
    api.post<EquivalenceGroupResponse>(`/projects/${projectId}/equivalence-groups/${groupId}/columns`, { column_ids: columnIds }).then(res => res.data),
  removeColumns: (projectId: number, groupId: number, columnIds: number[]) =>
    api.post<EquivalenceGroupRemoveColumnsResponse>(`/projects/${projectId}/equivalence-groups/${groupId}/columns/remove`, { column_ids: columnIds }).then(res => res.data),
  suggest: (projectId: number) =>
    api.get<EquivalenceSuggestResponse>(`/projects/${projectId}/equivalence-groups/suggest`).then(res => res.data),
  reorder: (projectId: number, groupIds: number[]) =>
    api.post(`/projects/${projectId}/equivalence-groups/reorder`, { group_ids: groupIds }).then(res => res.data),
  merge: (projectId: number, targetGroupId: number, sourceGroupId: number) =>
    api.post<EquivalenceGroupResponse>(`/projects/${projectId}/equivalence-groups/${targetGroupId}/merge/${sourceGroupId}`).then(res => res.data),
  findMatches: (projectId: number, columnIds: number[], minSimilarity?: number) =>
    api.post<FindMatchesResponse>(`/projects/${projectId}/equivalence-groups/find-matches`, { column_ids: columnIds, ...(minSimilarity != null ? { min_similarity: minSimilarity } : {}) }).then(res => res.data),
  /** Tier 3 crosswalk drag-to-swap. Atomically swaps the equivalence_group_id
   * of each pair. See backend routers/equivalence.py::swap_columns — three-phase
   * null-intermediate pattern with post-swap #290 validator (foot-gun).
   * Rate limited 60/minute. May raise structured 409/400 errors:
   *   { error: "type_mismatch" | "cross_dataset" | "not_linked" | "cross_dataset_unpaired", ... }
   */
  swap: (projectId: number, swaps: ColumnSwap[]) =>
    api.post<EquivalenceGroupSwapResponse>(`/projects/${projectId}/equivalence-groups/swap`, { swaps }).then(res => res.data),
}
